File: android/ac-markdown/src/main/kotlin/com/microsoft/adaptivecards/markdown/MarkdownRenderer.kt
package com.microsoft.adaptivecards.markdown

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.sp

/**
 * Renders markdown tokens to AnnotatedString for Compose Text views
 */
class MarkdownRenderer {
    
    companion object {
        /**
         * Convert markdown tokens to AnnotatedString
         * @param tokens The parsed markdown tokens
         * @param fontSize Base font size to use
         * @param color Base text color
         * @return AnnotatedString with markdown styling applied
         */
        fun render(
            tokens: List<MarkdownToken>,
            fontSize: TextUnit = 14.sp,
            color: Color = Color.Black
        ): AnnotatedString {
            return buildAnnotatedString {
                for (token in tokens) {
                    renderToken(token, fontSize, color)
                }
            }
        }
        
        private fun AnnotatedString.Builder.renderToken(
            token: MarkdownToken,
            fontSize: TextUnit,
            color: Color
        ) {
            when (token) {
                is MarkdownToken.Text -> {
                    pushStyle(SpanStyle(color = color, fontSize = fontSize))
                    append(token.text)
                    pop()
                }
                
                is MarkdownToken.Bold -> {
                    pushStyle(SpanStyle(
                        color = color,
                        fontSize = fontSize,
                        fontWeight = FontWeight.Bold
                    ))
                    append(token.text)
                    pop()
                }
                
                is MarkdownToken.Italic -> {
                    pushStyle(SpanStyle(
                        color = color,
                        fontSize = fontSize,
                        fontStyle = FontStyle.Italic
                    ))
                    append(token.text)
                    pop()
                }
                
                is MarkdownToken.Code -> {
                    pushStyle(SpanStyle(
                        color = color,
                        fontSize = fontSize,
                        fontFamily = FontFamily.Monospace,
                        background = Color.Gray.copy(alpha = 0.2f)
                    ))
                    append(token.text)
                    pop()
                }
                
                is MarkdownToken.Link -> {
                    pushStyle(SpanStyle(
                        color = Color.Blue,
                        fontSize = fontSize,
                        textDecoration = TextDecoration.Underline
                    ))
                    pushStringAnnotation(
                        tag = "URL",
                        annotation = token.url
                    )
                    append(token.text)
                    pop()
                    pop()
                }
                
                is MarkdownToken.Header -> {
                    val headerSize = when (token.level) {
                        1 -> fontSize * 2.0f
                        2 -> fontSize * 1.5f
                        3 -> fontSize * 1.25f
                        else -> fontSize
                    }
                    pushStyle(SpanStyle(
                        color = color,
                        fontSize = headerSize,
                        fontWeight = FontWeight.Bold
                    ))
                    append(token.text)
                    pop()
                }
                
                is MarkdownToken.BulletItem -> {
                    pushStyle(SpanStyle(color = color, fontSize = fontSize))
                    append("• ${token.text}")
                    pop()
                }
                
                is MarkdownToken.NumberedItem -> {
                    pushStyle(SpanStyle(color = color, fontSize = fontSize))
                    append("${token.number}. ${token.text}")
                    pop()
                }
                
                is MarkdownToken.LineBreak -> {
                    append("\n")
                }
            }
        }
    }
}

File: ios/Sources/ACMarkdown/MarkdownRenderer.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import SwiftUI

/// Renders markdown tokens to AttributedString for SwiftUI Text views
public class MarkdownRenderer {

    /// URL schemes considered safe for user navigation.
    ///
    /// Only these schemes are rendered as clickable links. All other schemes
    /// (e.g. `javascript:`, `data:`, `file:`, `vbscript:`, custom app schemes)
    /// are blocked to prevent XSS, phishing, and open-redirect attacks.
    ///
    /// - `http` / `https` — standard web URLs
    /// - `mailto` — email composition
    /// - `tel` — phone dialer
    ///
    /// See: GHSA-r5qq-54gp-7gcx
    private static let allowedSchemes: Set<String> = ["http", "https", "mailto", "tel"]

    /// Returns true if the URL uses a safe, allowed scheme.
    ///
    /// URLs without a scheme or with a disallowed scheme return false.
    public static func isSafeUrl(_ url: URL) -> Bool {
        guard let scheme = url.scheme?.lowercased() else { return false }
        return allowedSchemes.contains(scheme)
    }

    /// Convert markdown tokens to AttributedString
    /// - Parameters:
    ///   - tokens: The parsed markdown tokens
    ///   - font: Base font to use
    ///   - color: Base text color
    /// - Returns: AttributedString with markdown styling applied
    public static func render(
        tokens: [MarkdownToken],
        font: Font = .body,
        color: Color = .primary
    ) -> AttributedString {
        var result = AttributedString()

        for token in tokens {
            let attributed = renderToken(token, font: font, color: color)
            result.append(attributed)
        }

        return result
    }

    private static func renderToken(
        _ token: MarkdownToken,
        font: Font,
        color: Color
    ) -> AttributedString {
        switch token {
        case .text(let text):
            var attributed = AttributedString(text)
            attributed.font = font
            attributed.foregroundColor = color
            return attributed

        case .bold(let text):
            var attributed = AttributedString(text)
            attributed.font = font.bold()
            attributed.foregroundColor = color
            return attributed

        case .italic(let text):
            var attributed = AttributedString(text)
            attributed.font = font.italic()
            attributed.foregroundColor = color
            return attributed

        case .code(let text):
            var attributed = AttributedString(text)
            attributed.font = .system(.body, design: .monospaced)
            attributed.foregroundColor = color
            attributed.backgroundColor = Color.gray.opacity(0.2)
            return attributed

        case .link(let text, let url):
            var attributed = AttributedString(text)
            attributed.font = font
            if let parsedUrl = URL(string: url), Self.isSafeUrl(parsedUrl) {
                attributed.foregroundColor = .blue
                attributed.underlineStyle = .single
                attributed.link = parsedUrl
            } else {
                attributed.foregroundColor = color
            }
            return attributed

        case .header(let level, let text):
            // Use relative scaling from base font to match Android behavior
            // (fontSize * 2.0 for H1, * 1.5 for H2, * 1.25 for H3)
            let headerFont: Font = switch level {
            case 1: .system(size: 28, weight: .bold)
            case 2: .system(size: 21, weight: .bold)
            case 3: .system(size: 17.5, weight: .semibold)
            default: font.bold()
            }
            var attributed = AttributedString(text)
            attributed.font = headerFont
            attributed.foregroundColor = color
            return attributed

        case .bulletItem(let text):
            var result = AttributedString("• ")
            result.font = font
            result.foregroundColor = color
            for token in MarkdownParser.parse(text) {
                result.append(renderToken(token, font: font, color: color))
            }
            return result

        case .numberedItem(let number, let text):
            var result = AttributedString("\(number). ")
            result.font = font
            result.foregroundColor = color
            for token in MarkdownParser.parse(text) {
                result.append(renderToken(token, font: font, color: color))
            }
            return result

        case .lineBreak:
            return AttributedString("\n")
        }
    }
}
